src/runtime/profiler_common.cpp

```diff
@@ -229,39 +257,67 @@ WEAK int halide_profiler_pipeline_start(void *user_context,
+    LockProfiler lock(s);
+
+    if (instance->should_collect_statistics) {
+        halide_profiler_pipeline_stats *p = instance->pipeline_stats;
+
+        // Retire the instance, accumulating statistics onto the statistics for this
+        // pipeline. Fields related to memory usages are tracked in the pipeline stats
+        p->samples += instance->samples;
+        p->time += instance->time;
+        p->active_threads_numerator += instance->active_threads_numerator;
+        p->active_threads_denominator += instance->active_threads_denominator;
+        p->memory_total += instance->memory_total;
+        p->memory_peak = max(p->memory_peak, instance->memory_peak);
+        p->num_allocs += instance->num_allocs;
+        p->runs++;
+
+        for (int f = 0; f < p->num_funcs; f++) {
+            halide_profiler_func_stats *func = p->funcs + f;
+            const halide_profiler_func_stats *instance_func = instance->funcs + f;
+            func->time += instance_func->time;
+            func->active_threads_numerator += instance_func->active_threads_numerator;
+            func->active_threads_denominator += instance_func->active_threads_denominator;
+            func->num_allocs += instance_func->num_allocs;
+            func->stack_peak = max(func->stack_peak, instance_func->stack_peak);
+            func->memory_peak = max(func->memory_peak, instance_func->memory_peak);
+            func->memory_total += instance_func->memory_total;
+        }
     }
-    p->runs++;
 
-    return p->first_func_id;
+    // Remove myself from the doubly-linked list
+    *(instance->prev_next) = instance->next;
+    if (instance->next) {
+        instance->next->prev_next = instance->prev_next;
+    }
+    return 0;
 }
 
 WEAK void halide_profiler_stack_peak_update(void *user_context,
-                                            void *pipeline_state,
+                                            halide_profiler_instance_state *instance,
                                             uint64_t *f_values) {
-    halide_profiler_pipeline_stats *p_stats = (halide_profiler_pipeline_stats *)pipeline_state;
-    halide_abort_if_false(user_context, p_stats != nullptr);
-
     // Note: Update to the counter is done without grabbing the state's lock to
     // reduce lock contention. One potential issue is that other call that frees the
     // pipeline and function stats structs may be running in parallel. However, the
     // current desctructor (called on profiler shutdown) does not free the structs
     // unless user specifically calls halide_profiler_reset().
 
     // Update per-func memory stats
-    for (int i = 0; i < p_stats->num_funcs; ++i) {
+    for (int i = 0; i < instance->pipeline_stats->num_funcs; ++i) {
         if (f_values[i] != 0) {
-            sync_compare_max_and_swap(&(p_stats->funcs[i]).stack_peak, f_values[i]);
+            sync_compare_max_and_swap(&(instance->funcs[i]).stack_peak, f_values[i]);
```

Comment:
Just double checking: this one still needs to be atomic because of parallelism within the pipeline, right?

Reply:
Yes.